File: 211. Add and Search Word - Data structure design/Java/Solution.java
public class WordDictionary {
    TrieNode root;

    class TrieNode {
        TrieNode[] children = new TrieNode[26];
        String val = "";
    }

    /**
     * Initialize your data structure here.
     */
    public WordDictionary() {
        root = new TrieNode();
    }

    /**
     * Adds a word into the data structure.
     */
    public void addWord(String word) {
        TrieNode node = root;
        for (char letter : word.toCharArray()) {
            if (node.children[letter - 'a'] == null) {
                node.children[letter - 'a'] = new TrieNode();
            }
            node = node.children[letter - 'a'];
        }
        node.val = word;
    }

    /**
     * Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter.
     */
    public boolean search(String word) {
        class Utils {
            boolean search(String word, int index, TrieNode node) {
                if (index == word.length()) {
                    return !node.val.isEmpty();
                }
                if (word.charAt(index) == '.') {
                    for (TrieNode child : node.children) {
                        if (child != null) {
                            if (search(word, index + 1, child)) {
                                return true;
                            }
                        }
                    }
                } else {
                    TrieNode child = node.children[word.charAt(index) - 'a'];
                    return child != null && search(word, index + 1, child);
                }
                return false;
            }
        }
        Utils utils = new Utils();
        return utils.search(word, 0, root);
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */

File: 211. Add and Search Word - Data structure design/Python/Solution.py
class TrieNode:
    def __init__(self):
        self.children = [None] * 26
        self.val = None


class WordDictionary(object):

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = TrieNode()

    def addWord(self, word):
        """
        Adds a word into the data structure.
        :type word: str
        :rtype: void
        """
        node = self.root
        for letter in word:
            if not node.children[ord(letter) - ord('a')]:
                node.children[ord(letter) - ord('a')] = TrieNode()
            node = node.children[ord(letter) - ord('a')]
        node.val = word

    def search(self, word):
        """
        Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter.
        :type word: str
        :rtype: bool
        """

        def search(word, index, node):
            if index == len(word):
                return node.val is not None
            if word[index] == '.':
                for child in node.children:
                    if child:
                        if (search(word, index + 1, child)):
                            return True
            else:
                child = node.children[ord(word[index]) - ord('a')]
                return child is not None and search(word, index + 1, child)
            return False

        return search(word, 0, self.root)


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)
